src/server/pfs/server/master.go

```diff
@@ -87,78 +93,134 @@ func (d *driver) master(ctx context.Context) {
 					return nil
+				}), "for each repo")
+			}, dbutil.WithReadOnly()); err != nil {
+				return errors.Wrap(err, "list repos")
+			}
+			for _, event := range existingRepos {
+				if err := d.manageRepos(ctx, ring, repos, event); err != nil {
+					return errors.Wrap(err, "manage repos for existing entries")
 				}
-				ctx, cancel := pctx.WithCancel(ctx)
-				repos[key] = cancel
-				go func() {
-					backoff.RetryUntilCancel(ctx, func() (retErr error) { //nolint:errcheck
-						ctx, cancel := pctx.WithCancel(ctx)
-						defer cancel()
-						var err error
-						ctx, err = ring.Lock(ctx, lockPrefix)
-						if err != nil {
-							return errors.Wrap(err, "error locking repo lock")
-						}
-						defer errors.Invoke1(&retErr, ring.Unlock, lockPrefix, "unlocking repo lock")
-						var eg errgroup.Group
-						eg.Go(func() error {
-							return backoff.RetryUntilCancel(ctx, func() error {
-								return d.manageBranches(ctx, key)
-							}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-								log.Error(ctx, "error managing branches", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-								return nil
-							})
-						})
-						eg.Go(func() error {
-							return backoff.RetryUntilCancel(ctx, func() error {
-								return d.finishRepoCommits(ctx, key)
-							}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-								log.Error(ctx, "error finishing repo commits", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-								return nil
-							})
-						})
-						return errors.EnsureStack(eg.Wait())
-					}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-						log.Error(ctx, "error managing repo", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-						return nil
-					})
-				}()
-				return nil
-			})
-			return errors.EnsureStack(err)
+			}
+			return errors.Wrap(eg.Wait(), "waiting for manageRepos goroutines to finish")
 		},
 	)
 }
 
+func (d *driver) manageRepos(ctx context.Context, ring *consistenthashing.Ring, repos map[pfsdb.RepoID]context.CancelFunc, ev *postgres.Event) error {
+	if ev.Err != nil {
+		return ev.Err
+	}
+	eventID := pfsdb.RepoID(ev.Id)
+	lockPrefix := path.Join("repos", fmt.Sprintf("%d", ev.Id))
+	if ev.Type == postgres.EventDelete {
+		if cancel, ok := repos[eventID]; ok {
+			if err := ring.Unlock(lockPrefix); err != nil {
+				return err
+			}
+			cancel()
+			delete(repos, eventID)
+		}
+		return nil
+	}
+	if _, ok := repos[eventID]; ok {
+		return nil
+	}
+	var repo *pfs.RepoInfo
+	var err error
+	if err := dbutil.WithTx(ctx, d.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+		repo, err = pfsdb.GetRepo(ctx, tx, eventID)
+		return errors.Wrap(err, "get repo from event id")
+	}, dbutil.WithReadOnly()); err != nil {
+		return errors.Wrap(err, "get repo")
+	}
+	key := pfsdb.RepoKey(repo.Repo)
+	ctx, cancel := pctx.WithCancel(ctx)
+	repos[eventID] = cancel
```

Comment:
This is a concurrent write to the map.

Reply:
Interesting! I guess we were doing this before, but maybe we never ran into errors because of WatchF? (maybe watchF gets a lock under the hood)